### ISSUE TITLE
error BatchNormalization

### ISSUE THREAD

**Issue by MLTOL** [Microsoft/CNTK#2164]
I have converted my converted my NDLNetworkBuilder to BrainscriptNetworkBuilder following the steps on: https://docs.microsoft.com/en-us/cognitive-toolkit/BrainScript-Network-Builder. I have changed my .ndl file to a .bs file and I use a Macros file called Macros.bs.

> 
      bn = BatchNormalization(t, sc, b, m, isd, eval = false, spatial = false, normalizationTimeConstant = bnTimeConst)
                             0^
EXCEPTION

 occurred. while evaluating: Macros.bs(42): function expects 6 positional parameters, 5 were provided

For the NDLNetworkBuilder this function has worked fine for me, but I need the conversion because it does not work with 3D convolution.

Are there any suggestions why this function throws this error? Wat do they mean with positional parameters?

**Comment by cha-zhang** [comment]
Which network do you use? Have you tried Macros.bs like this one: 
https://github.com/Microsoft/CNTK/blob/master/Examples/Image/Classification/ResNet/BrainScript/Macros.bs

**Comment by MLTOL** [comment]
Hi, thank you for your answer. I use a convolutional neural network. I call the functions:

`conv = ConvNDBNReLULayer(featuresCTA, cMap1, 75, kW1, kH1, inMap, hStride1, vStride1, convWScale, convBValue, scValue, bnTimeConst)`

this will use the functions in my Macros.bs file:

`ConvNDBNReLULayer(inp, outMap, inWCount, kW, kH, inMap, hStride, vStride, wScale, bValue, scValue, bnTimeConst) = {
c = ConvNDBNLayer(inp, outMap, inWCount, kW, kH, inMap, hStride, vStride, wScale, bValue, 
scValue, bnTimeConst)
y = RectifiedLinear(c)
}.y

ConvNDBNLayer(inp, outMap, inWCount, kW, kH, inMap, hStride, vStride, wScale, bValue, scValue, bnTimeConst) = {
W = ParameterTensor{outMap, inWCount, init=Gaussian, initValueScale=wScale, initOnCPUOnly=true}
c = ConvNDBNLayerW(W, inp, outMap, kW, kH, inMap, hStride, vStride, bValue, scValue, 
          bnTimeConst)
}.c

ConvNDBNLayerW(W, inp, outMap, kW, kH, inMap, hStride, vStride, bValue, scValue, bnTimeConst) = {
b = ParameterTensor{outMap, 1, init=fixedValue, value=bValue}
sc = ParameterTensor{outMap, 1, init=fixedValue, value=scValue}
m = ParameterTensor{outMap, 1, init=fixedValue, value=0, learningRateMultiplier=0}
isd = ParameterTensor{outMap, 1, init=fixedValue, value=0, learningRateMultiplier=0}

c=Convolution(w, inp, (kW:kH:inMap), mapCount=outMap, stride=(hStride:vStride:inMap), sharing=
         (true:true:true), autoPadding=(true:true:false), lowerPad=0, upperPad=0) 
y = BatchNormalization(c, sc, b, m, isd, eval=false, spatial=true, 
          normalizationTimeConstant=bnTimeConst)
}.y

`

**Comment by cha-zhang** [comment]
The layer reference of BrainScript is available here: 
https://docs.microsoft.com/en-us/cognitive-toolkit/BrainScript-Layers-Reference 

BatchNormalizationLayer {spatialRank = 0,
                         normalizationTimeConstant = 5000,
                         initialScale = 1, epsilon = 0.00001, useCntkEngine = true}

It is very different from NDL and you should probably rewrite your Macro.bs.

**Comment by cha-zhang** [comment]
And here's the low level API: 
https://docs.microsoft.com/en-us/cognitive-toolkit/BatchNormalization 

BatchNormalization(input, scale, bias, runMean, runVariance, spatial, normalizationTimeConstant = 0, blendTimeConstant = 0, epsilon = 0.00001, useCntkEngine = true)

**Comment by cha-zhang** [comment]
Closing this issue due to lack of response. Please reopen if needed.
